django/db/models/expressions.py

```diff
@@ -1718,6 +1718,14 @@ def as_sql(
         else:
             default_sql, default_params = compiler.compile(self.default)
         if not case_parts:
+            from django.db.models.functions import Cast
+
+            if isinstance(self.default, Value) and not extra_context.get(
+                "no_cast_default_case"
+            ):
+                default_sql, default_params = compiler.compile(
+                    Cast(self.default, self.default.output_field)
+                )
```

Comment:
Thinking more of it, `default_sql` can also be originating from a `case` (see line 1714) so using `self.default` systematically is flawed. I think we should instead adjust the logic about to defer compilation and do

```python
for case in self.cases:
    try:
        case_sql, case_params = compiler.compile(case)
    except EmptyResultSet:
        continue
    except FullResultSet:
        default = case.result
        break
    case_parts.append(case_sql)
    sql_params.extend(case_params)
else:
    default = self.default
if not case_parts:
    from django.db.models.functions import Cast

    if isinstance(default, Value):
        default = Cast(default, default.output_field)
    default_sql, default_params = compiler.compile(self.default)
```